### PR TITLE
Lookup limit push down

### PR DESCRIPTION
**Pull request by bright-starry-sky**

### CODE REVIEW

File: src/interface/storage.thrift [hunk 1]
```diff
@@ -521,6 +520,8 @@ struct LookupIndexRequest {
     // The list of property names. Should not be empty.
     // Support kVid and kTag for vertex, kSrc, kType, kRank and kDst for edge.
     4: optional list<binary>                return_columns,
+    // max row count of each partition in this response, 0 means no limit.
+    5: i64                                  limit = 0,
```
Comment (Shylock-Hg) [inline src/interface/storage.thrift:524]: Try `optional` and `NULL` means no limit instead zero.

Comment (Shylock-Hg) [reply src/interface/storage.thrift:524]: Or you could try max of i64

Comment (Shylock-Hg) [reply src/interface/storage.thrift:524]: Or negative value for this.

**Comment by Shylock-Hg** [comment]
And you should add a test case.

File: src/interface/storage.thrift (hunk 1 above)
Comment (bright-starry-sky) [reply src/interface/storage.thrift]: I don't see any advantage in your approach. your approach may make the code more complex.

Comment (Shylock-Hg) [reply src/interface/storage.thrift]: The limit 0 is allowed.

Comment (Shylock-Hg) [reply src/interface/storage.thrift]: And the other limit field in request is optional too.

Comment (bright-starry-sky) [reply src/interface/storage.thrift]: > And the other limit field in request is optional too.

good point.

**Comment by bright-starry-sky** [comment]
Addressed comments.

**Comment by bright-starry-sky** [comment]
Related #2602

File: src/storage/index/LookupBaseProcessor.h [hunk 2]
```diff
@@ -81,6 +81,7 @@ class LookupBaseProcessor : public BaseProcessor<RESP> {
   // Save schemas when column is out of index, need to read from data
   std::vector<std::shared_ptr<const meta::NebulaSchemaProvider>> schemas_;
   std::vector<size_t> deDupColPos_;
+  int64_t limit_ = -1;
```
Comment (panda-sheep) [inline src/storage/index/LookupBaseProcessor.h:84]: maybe the default value of `limit `is 0 (0 means no limit) better?
This is not important

**Review by panda-sheep — APPROVED** [review]
Good job！

**Comment by bright-starry-sky** [comment]
Corrected comments. thanks @panda-sheep

**Review by panda-sheep — DISMISSED** [review]

**Review by critical27 — APPROVED** [review]
Generally LGTM, wait graph PR is ready.

**Review by panda-sheep — APPROVED** [review]

**Comment by bright-starry-sky** [comment]
related #2823

**Comment by cangfengzhs** [comment]
The `limit` must be the top node of the execution plan, otherwise the size of the final dataset will be smaller than the limit count due to the filter (and possibly dedup) discarding part of the data.

**Comment by Shylock-Hg** [comment]
> The `limit` must be the top node of the execution plan, otherwise the size of the final dataset will be smaller than the limit count due to the filter (and possibly dedup) discarding part of the data.

Yes the limit push down can't over the filter operation.

**Comment by bright-starry-sky** [comment]
> The `limit` must be the top node of the execution plan, otherwise the size of the final dataset will be smaller than the limit count due to the filter (and possibly dedup) discarding part of the data.

Good point. Thanks for your remind. 
The limit can't cover filter, ttl and dedup.  I'll deal with it later.

**Pull request merged**